compiler/z/codegen/OMRTreeEvaluator.cpp

```diff
@@ -9672,25 +9675,50 @@ OMR::Z::TreeEvaluator::loadaddrEvaluator(TR::Node * node, TR::CodeGenerator * cg
                cursor = generateRILInstruction(cg, TR::InstOpCode::LARL, node, targetRegister, node->getSymbolReference(),
                                                 reinterpret_cast<void*>(sym->getStaticAddress()),  NULL);
                }
+            else if (cg->needRelocationsForBodyInfoData() && sym && sym->isRecompilationCounter())
+               {
+               cursor = generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister,
+                                                     (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(),
+                                                     TR_BodyInfoAddress, NULL, NULL, NULL);
+               }
+            else if (cg->needClassAndMethodPointerRelocations() && node->getSymbol()->isClassObject())
+               {
+               cursor = generateRegLitRefInstruction(cg, TR::InstOpCode::getLoadOpCode(), node, targetRegister,
+                                                     (uintptr_t) node->getSymbol()->getStaticSymbol()->getStaticAddress(),
+                                                     TR_ClassAddress, NULL, NULL, NULL);
+               }
+            else if (cg->needRelocationsForStatics() && sym)
+               {
+               int32_t reloType = 0;
+               if (sym->isStartPC())
```

Comment:
Is this path ever going to be taken? I see that we already have a case for `if (sym && sym->isStartPC())` on line 9673

Reply:
I'm not sure why this was there or why this worked for AOT because the address should be relocated for those compiles. I haven't been able to see failures related to this. I've moved the above check below. See edda716f5b4f9b4e4d46d06fe1dd3eae2095f047 for details. Now we will capture the AOT/JITServer cases before the default path on 9673. Functionally this should get us the same behaviour as before.

Reply:
Actually I think what I said above is incorrect. So I'll back out my commit. @dsouzai FYI

Reply:
In AOT compilations, we are never able to get our hands on the start PC of another method; we pretend like all other methods are interpreted so that we always generate an interpreter dispatch snippet. The reason for this is because we don't know what order methods are going to be loaded in, so it's better for the interpreter dispatch snippet to just patch the call when the callee is compiled/loaded.

Reply:
Assuming the code at line 9673 is correct, we can delete this new code that sets `reloType = TR_AbsoluteMethodAddress;`

Reply:
Considering that the code above is using an `LARL` instruction, it may not be correct for JITServer. But I think the address that it's trying to materialize would be a client address so it doesn't need a relocation. So we should be able to use something similar to the following approach to materialize the address into a register:

```
//disable LARL for AOT for easier relocation
   //cannot safely generate LARL for longs on 31-bit
   else if (cg->canUseRelativeLongInstructions(value) &&
            TR::Compiler->target.is64Bit() && !(comp->compileRelocatableCode()))
      {
      cursor = generateRILInstruction(cg, TR::InstOpCode::LARL, node, targetRegister, reinterpret_cast<void*>(value), cursor);
      }
   else
      {
      int32_t high32 = value >> 32;

      cursor = generateRILInstruction(cg, TR::InstOpCode::LLIHF, node, targetRegister, high32, cursor);
      cursor = generateRILInstruction(cg, TR::InstOpCode::IILF, node, targetRegister, static_cast<int32_t>(value), cursor);
      }
```

Reply:
e5ed866ee8877bbc0262a363a3e9647a24c2724d

Reply:
I would prefer the dead code for startPC at line 9702 be deleted:
```
              if (sym->isStartPC())
                 {
                  reloType = TR_AbsoluteMethodAddress;
                  }
```

Reply:
I've deleted the code here: 70e12485427fce939440d385ff0461712b8a893b

I've also removed the code suggested in https://github.com/eclipse/omr/pull/4458#discussion_r454461727 and reverted back to the original `LARL` implementation. It looks like this approach works for JITServer already. The reason being that in this case, we only seem to be seeing a `startPC` that points to the `startPC` of the method being compiled currently. Since `LARL` is a position dependent instruction, we can calculate the offset needed to jump back to the `startPC` on the server side and don't need any further relocation. So in this case, we can indeed use `LARL` for a JITServer compile.

Secondly, materializing the address of the `startPC` into a register using the `LLIHF` and `IILF` would be incorrect without a relocation. This is because if the `startPC` only points to the method itself, then that address is not available to us until the code cache is allocated on the client side - something not done when performing the compile on the server.

During binaryEncoding of the `LARL` instruction, the offset is calculated for cases where the `startPC` is the method itself, and relocations are added if needed (ex. if the method being called is a helper).

A similar problem to the one described above was encountered on Power as well. See https://github.com/eclipse/openj9/blob/2eba29ba5bb1b964e65420a8fa03116965a99a15/runtime/compiler/p/codegen/J9PPCInstruction.cpp#L140-L160 for more context.

Reply:
Note: I've also removed the code suggested in https://github.com/eclipse/omr/pull/4458#discussion_r454461727 and reverted back to the original `LARL` implementation. It looks like this approach works for JITServer already. The reason being that in this case, we only seem to be seeing a `startPC` that points to the `startPC` of the method being compiled currently. Since `LARL` is a position dependent instruction, we can calculate the offset needed to jump back to the `startPC` on the server side and don't need any further relocation. So in this case, we can indeed use `LARL` for a JITServer compile.

Secondly, materializing the address of the `startPC` into a register using the `LLIHF` and `IILF` would be incorrect without a relocation. This is because if the `startPC` only points to the method itself, then that address is not available to us until the code cache is allocated on the client side - something not done when performing the compile on the server.

During binaryEncoding of the `LARL` instruction, the offset is calculated for cases where the `startPC` is the method itself, and relocations are added if needed (ex. if the method being called is a helper).

A similar problem to the one described above was encountered on Power as well. See https://github.com/eclipse/openj9/blob/2eba29ba5bb1b964e65420a8fa03116965a99a15/runtime/compiler/p/codegen/J9PPCInstruction.cpp#L140-L160 for more context.